core/lib/zksync_core/src/sync_layer/snapshots.rs

```diff
@@ -0,0 +1,496 @@
+use crate::state_keeper::{L1BatchExecutorBuilder, MainBatchExecutorBuilder, ZkSyncStateKeeper};
+use crate::sync_layer::fetcher::MainNodeFetcherCursor;
+use crate::sync_layer::snapshots::SnapshotApplierError::*;
+use crate::sync_layer::{ActionQueue, ExternalIO, MainNodeClient, SyncState};
+use anyhow::Context;
+use multivm::vm_1_3_2::zk_evm_1_3_3::ethereum_types::Address;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::path::Path;
+use tokio::sync::watch;
+use zksync_dal::connection::DbVariant;
+use zksync_dal::{ConnectionPool, StorageProcessor};
+use zksync_merkle_tree::recovery::{MerkleTreeRecovery, RecoveryEntry};
+use zksync_merkle_tree::RocksDBWrapper;
+use zksync_object_store::{ObjectStore, ObjectStoreFactory};
+use zksync_types::block::{BlockGasCount, MiniblockHeader};
+use zksync_types::snapshots::{
+    AppliedSnapshotStatus, SnapshotFactoryDependencies, SnapshotHeader, SnapshotStorageLog,
+    SnapshotStorageLogsChunk, SnapshotStorageLogsStorageKey,
+};
+use zksync_types::{
+    L1BatchNumber, L2ChainId, ProtocolVersionId, StorageKey, StorageLog, StorageLogKind, H256,
+};
+
+pub struct StateKeeperConfig {
+    pub connection_pool: ConnectionPool,
+    pub state_keeper_db_path: String,
+    pub l2_erc20_bridge_addr: Address,
+    pub chain_id: L2ChainId,
+    pub main_node_url: String,
+    pub enum_index_migration_chunk_size: usize,
+}
+pub struct SnapshotApplier<'a, 'b, 'd> {
+    connection_pool: &'a ConnectionPool,
+    client: &'a dyn MainNodeClient,
+    recovery: MerkleTreeRecovery<'b, RocksDBWrapper>,
+    blob_store: Box<dyn ObjectStore>,
+    applied_snapshot_status: AppliedSnapshotStatus,
+    snapshot: SnapshotHeader,
+    state_keeper_config: StateKeeperConfig,
+    metadata_calculator_config: MetadataCalculatorConfig<'d>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum SnapshotApplierError {
+    #[error("canceled")]
+    Canceled(String),
+    #[error(transparent)]
+    Fatal(#[from] anyhow::Error),
+    #[error("retryable")]
+    Retryable(String),
+}
+
+impl<'a, 'b, 'd> SnapshotApplier<'a, 'b, 'd> {
+    pub async fn new(
+        connection_pool: &'a ConnectionPool,
+        client: &'a dyn MainNodeClient,
+        merkle_tree_db_path: &String,
+        state_keeper_config: StateKeeperConfig,
+        metadata_calculator_config: MetadataCalculatorConfig<'d>,
+    ) -> anyhow::Result<SnapshotApplier<'a, 'b, 'd>, SnapshotApplierError> {
+        let mut storage = connection_pool
+            .access_storage_tagged("snapshots_applier")
+            .await
+            .unwrap();
+        let mut storage = storage.start_transaction().await.unwrap();
+
+        let mut applied_snapshot_status = storage
+            .applied_snapshot_status_dal()
+            .get_applied_snapshot_status()
+            .await
+            .unwrap();
+
+        if !storage.blocks_dal().is_genesis_needed().await.unwrap()
+            && applied_snapshot_status.is_none()
+        {
+            return Err(Canceled(
```

Comment:
Bikeshedding: Personally, importing enum variants makes code significantly more difficult to read. There are exceptions (short functions / methods mostly consisting of a `match`), but they only prove the rule.